src/Analyzer.JsonRuleEngine/JsonRuleEngine.cs

```diff
@@ -33,33 +34,39 @@ public JsonRuleEngine(IJsonLineNumberResolver lineNumberResolver)
         /// <param name="templateContext">The template context to evaluate.</param>
         /// <param name="ruleDefinitions">The JSON rules to evaluate the template with.</param>
         /// <returns>The results of the rules against the template.</returns>
-        public IEnumerable<IResult> EvaluateRules(TemplateContext templateContext, string ruleDefinitions)
+        public IEnumerable<IEvaluation> EvaluateRules(TemplateContext templateContext, string ruleDefinitions)
         {
             List<RuleDefinition> rules = JsonConvert.DeserializeObject<List<RuleDefinition>>(ruleDefinitions);
 
             foreach(RuleDefinition rule in rules)
             {
                 var ruleExpression = rule.Evaluation.ToExpression();
-                var ruleResults = ruleExpression.Evaluate(
+                var evaluation = ruleExpression.Evaluate(
                     new JsonPathResolver(
                         templateContext.ExpandedTemplate,
                         templateContext.ExpandedTemplate.Path));
 
-                foreach (var result in ruleResults)
+                // If there are no matching cases of the rule, do not create an evaluation
+                if (!(evaluation as IEvaluation).HasResults())
```

Comment:
This is here because it was creating empty evaluations for rules where even the resource type was not present. Without this, there will always be one evaluation result for each rule even if the rule doesn't apply to the template.

Reply:
That makes sense.  But I think there's still a problem in the allOf scenario - allOf will never have results.